File: ios/Sources/WishpoolCore/AgentApi.swift
import Foundation

/// AI Server 调用接口 — 对标 Android 端 AgentApi.kt
/// 调用 AI Server，失败时降级到本地模板
public actor AgentApi {
    private let aiServerUrl: String

    public init(aiServerUrl: String = "http://localhost:3100") {
        self.aiServerUrl = aiServerUrl
    }

    /// 调用 AI Server 生成方案
    public func generatePlan(wishInput: String) async -> AiPlanResult {
        do {
            guard let url = URL(string: "\(aiServerUrl)/plan") else {
                return .success(plan: generateLocalTemplate(wishInput: wishInput), provider: "local-fallback")
            }

            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.timeoutInterval = 15

            let body: [String: Any] = ["wishInput": wishInput]
            request.httpBody = try JSONSerialization.data(withJSONObject: body)

            let (data, _) = try await URLSession.shared.data(for: request)
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]

            if json["success"] as? Bool == true, let planJson = json["plan"] as? [String: Any] {
                let plan = parsePlan(planJson, fallbackWishText: wishInput)
                let provider = json["provider"] as? String ?? "ai-server"
                return .success(plan: plan, provider: provider)
            } else {
                return .success(plan: generateLocalTemplate(wishInput: wishInput), provider: "local-fallback")
            }
        } catch {
            return .success(plan: generateLocalTemplate(wishInput: wishInput), provider: "local-fallback")
        }
    }

    // MARK: - 解析

    private func parsePlan(_ json: [String: Any], fallbackWishText: String) -> GeneratedPlan {
        let optionsArray = json["decisionOptions"] as? [[String: Any]] ?? []
        let options = optionsArray.map { opt in
            DecisionOption(
                key: opt["key"] as? String ?? "",
                label: opt["label"] as? String ?? ""
            )
        }

        let stepsArray = json["planSteps"] as? [[String: Any]] ?? []
        let steps = stepsArray.map { s in
            AiPlanStep(
                num: s["num"] as? String ?? "①",
                title: s["title"] as? String ?? "",
                type: s["type"] as? String ?? "",
                typeColor: s["typeColor"] as? String ?? "#4AADA0",
                desc: s["desc"] as? String ?? ""
            )
        }

        return GeneratedPlan(
            wishText: json["wishText"] as? String ?? fallbackWishText,
            durationText: json["durationText"] as? String ?? "预计 5 天完成",
            decisionTitle: json["decisionTitle"] as? String ?? "",
            decisionOptions: options,
            planSteps: steps,
            category: json["category"] as? String ?? "生活体验",
            difficulty: json["difficulty"] as? String ?? "medium",
            estimatedDays: json["estimatedDays"] as? Int ?? 5
        )
    }
}

public enum AiPlanResult: Sendable {
    case success(plan: GeneratedPlan, provider: String)
    case error(message: String)
}

// MARK: - 本地模板降级方案

public func generateLocalTemplate(wishInput: String) -> GeneratedPlan {
    let input = wishInput.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()

    if input.containsAny("海边", "海滩", "放松", "旅行", "度假") {
        return GeneratedPlan(
            wishText: wishInput.trimmingCharacters(in: .whitespacesAndNewlines),
            durationText: "预计 3-5 天完成",
            decisionTitle: "AI 需要你决定：这次海边之行你更想要什么体验？",
            decisionOptions: [
                DecisionOption(key: "relax", label: "纯放松休闲"),
                DecisionOption(key: "activity", label: "海上活动体验"),
                DecisionOption(key: "photo", label: "拍照打卡风景"),
            ],
            planSteps: [
                AiPlanStep(num: "①", title: "筛选适合的海边目的地和住宿", type: "线上直出", typeColor: "#4AADA0", desc: "AI 自动搜索推荐"),
                AiPlanStep(num: "②", title: "预订交通和酒店，准备物品清单", type: "资源助力", typeColor: "#F5C842", desc: "平台资源助力"),
                AiPlanStep(num: "③", title: "找同行伙伴或当地向导推荐", type: "人群助力", typeColor: "#C084FC", desc: "AI匹配志同道合的旅友"),
                AiPlanStep(num: "④", title: "开始海边放松之旅", type: "需你到场", typeColor: "#F97316", desc: "你本人享受旅程"),
            ],
            category: "生活体验",
            difficulty: "easy",
            estimatedDays: 4
        )
    }

    if input.containsAny("学习", "技能", "考试", "读书") {
        return GeneratedPlan(
            wishText: wishInput.trimmingCharacters(in: .whitespacesAndNewlines),
            durationText: "预计 10-14 天完成",
            decisionTitle: "AI 需要你决定：你更偏好哪种学习模式？",
            decisionOptions: [
                DecisionOption(key: "self", label: "自主学习"),
                DecisionOption(key: "course", label: "系统课程"),
                DecisionOption(key: "mentor", label: "导师指导"),
            ],
            planSteps: [
                AiPlanStep(num: "①", title: "制定学习计划和时间安排", type: "线上直出", typeColor: "#4AADA0", desc: "AI 自动规划"),
                AiPlanStep(num: "②", title: "整理学习资源和材料", type: "资源助力", typeColor: "#F5C842", desc: "平台资源助力"),
                AiPlanStep(num: "③", title: "寻找学习伙伴或导师", type: "人群助力", typeColor: "#C084FC", desc: "AI匹配学习搭子"),
                AiPlanStep(num: "④", title: "开始系统化学习", type: "需你到场", typeColor: "#F97316", desc: "你本人投入学习"),
            ],
            category: "学习成长",
            difficulty: "medium",
            estimatedDays: 12
        )
    }

    if input.containsAny("运动", "跑步", "健身", "锻炼", "夜跑") {
        return GeneratedPlan(
            wishText: wishInput.trimmingCharacters(in: .whitespacesAndNewlines),
            durationText: "预计 7 天完成",
            decisionTitle: "AI 需要你决定：你更喜欢哪种运动节奏？",
            decisionOptions: [
                DecisionOption(key: "light", label: "轻松入门"),
                DecisionOption(key: "regular", label: "规律训练"),
                DecisionOption(key: "intense", label: "高强度挑战"),
            ],
            planSteps: [
                AiPlanStep(num: "①", title: "制定适合的运动计划", type: "线上直出", typeColor: "#4AADA0", desc: "AI 个性化规划"),
                AiPlanStep(num: "②", title: "准备运动装备和场地", type: "资源助力", typeColor: "#F5C842", desc: "平台资源助力"),
                AiPlanStep(num: "③", title: "寻找运动伙伴或教练", type: "人群助力", typeColor: "#C084FC", desc: "AI匹配运动搭子"),
                AiPlanStep(num: "④", title: "开始规律运动", type: "需你到场", typeColor: "#F97316", desc: "你本人坚持锻炼"),
            ],
            category: "运动健康",
            difficulty: "easy",
            estimatedDays: 7
        )
    }

    if input.containsAny("滑雪", "雪场", "单板", "双板") {
        return GeneratedPlan(
            wishText: wishInput.trimmingCharacters(in: .whitespacesAndNewlines),
            durationText: "预计 5 天完成",
            decisionTitle: "AI 需要你决定：你更偏好哪种滑雪体验？",
            decisionOptions: [
                DecisionOption(key: "beginner", label: "新手入门"),
                DecisionOption(key: "intermediate", label: "进阶挑战"),
                DecisionOption(key: "group", label: "组队同行"),
            ],
            planSteps: [
                AiPlanStep(num: "①", title: "筛选雪场 + 锁定新手友好线路", type: "线上直出", typeColor: "#4AADA0", desc: "AI 自动完成"),
                AiPlanStep(num: "②", title: "整理拼车时间 + 雪具租赁建议", type: "资源助力", typeColor: "#F5C842", desc: "平台资源助力"),
                AiPlanStep(num: "③", title: "匹配有车且节奏合适的滑雪搭子", type: "人群助力", typeColor: "#C084FC", desc: "AI发邀约·按滑雪画像"),
                AiPlanStep(num: "④", title: "按约定出发滑雪 + 回填体验反馈", type: "需你到场", typeColor: "#F97316", desc: "你本人参与"),
            ],
            category: "生活体验",
            difficulty: "medium",
            estimatedDays: 5
        )
    }

    // 通用模板
    return GeneratedPlan(
        wishText: wishInput.trimmingCharacters(in: .whitespacesAndNewlines),
        durationText: "预计 5 天完成",
        decisionTitle: "AI 需要你决定：你更偏好哪种执行方式？",
        decisionOptions: [
            DecisionOption(key: "solo", label: "独自完成"),
            DecisionOption(key: "partner", label: "寻找搭子"),
            DecisionOption(key: "community", label: "社区协助"),
        ],
        planSteps: [
            AiPlanStep(num: "①", title: "分析需求并制定计划", type: "线上直出", typeColor: "#4AADA0", desc: "AI 智能分析"),
            AiPlanStep(num: "②", title: "准备必要的资源和信息", type: "资源助力", typeColor: "#F5C842", desc: "平台资源助力"),
            AiPlanStep(num: "③", title: "寻找合适的协助伙伴", type: "人群助力", typeColor: "#C084FC", desc: "AI匹配合适搭子"),
            AiPlanStep(num: "④", title: "开始执行你的心愿", type: "需你到场", typeColor: "#F97316", desc: "你本人参与"),
        ],
        category: "生活体验",
        difficulty: "medium",
        estimatedDays: 5
    )
}

private extension String {
    func containsAny(_ keywords: String...) -> Bool {
        keywords.contains { self.contains($0) }
    }
}

File: android/app/src/main/java/com/wishpool/app/data/remote/AgentApi.kt
package com.wishpool.app.data.remote

import android.util.Log
import com.wishpool.app.domain.model.AiPlanStep
import com.wishpool.app.domain.model.DecisionOption
import com.wishpool.app.domain.model.GeneratedPlan
import com.wishpool.app.domain.model.WishAnalysis
import org.json.JSONObject

/**
 * AI Server 调用接口 — 对标 Web 端 agent-api.ts
 * 调用本地 AI Server (localhost:3100)，失败时降级到本地模板
 */
class AgentApi(
    private val aiServerUrl: String = AI_SERVER_URL,
) {
    private val http = com.wishpool.app.core.network.HttpClient(
        enableVerboseLogs = true,
    )

    /**
     * 调用 AI Server 生成方案
     * POST /plan { wishInput }
     */
    fun generatePlan(wishInput: String): AiPlanResult {
        return try {
            Log.d(TAG, "调用 AI Server 生成方案: $wishInput")
            val body = JSONObject()
                .put("wishInput", wishInput)

            val json = http.post("$aiServerUrl/plan", body.toString())
            val data = JSONObject(json)

            if (data.optBoolean("success", false)) {
                // 优先从 plan 字段解析（AI Server 返回的 GeneratedPlan 格式）
                val planJson = data.optJSONObject("plan")
                if (planJson != null) {
                    val plan = parsePlan(planJson, wishInput)
                    val provider = data.optString("provider", "ai-server")
                    Log.d(TAG, "AI 方案生成成功 ($provider)")
                    AiPlanResult.Success(plan, provider)
                } else {
                    Log.w(TAG, "AI Server 返回成功但无 plan 字段，降级到本地模板")
                    AiPlanResult.Success(generateLocalTemplate(wishInput), "local-template")
                }
            } else {
                val error = data.optString("error", "未知错误")
                Log.w(TAG, "AI Server 返回失败: $error，降级到本地模板")
                AiPlanResult.Success(generateLocalTemplate(wishInput), "local-fallback")
            }
        } catch (e: Exception) {
            Log.w(TAG, "AI Server 不可用，降级到本地模板: ${e.message}")
            AiPlanResult.Success(generateLocalTemplate(wishInput), "local-fallback")
        }
    }

    /**
     * 调用 AI Server 进行意图分析
     * POST /analyze { wish }
     */
    fun analyzeWish(wish: String): WishAnalysis? {
        return try {
            val body = JSONObject()
                .put("wish", wish)

            val json = http.post("$aiServerUrl/analyze", body.toString())
            val data = JSONObject(json)

            if (data.optBoolean("success", false)) {
                val analysis = data.optJSONObject("analysis") ?: return null
                parseAnalysis(analysis)
            } else {
                null
            }
        } catch (e: Exception) {
            Log.w(TAG, "AI Server 意图分析不可用: ${e.message}")
            null
        }
    }

    private fun parsePlan(json: JSONObject, fallbackWishText: String): GeneratedPlan {
        val options = buildList {
            val arr = json.optJSONArray("decisionOptions")
            if (arr != null) {
                for (i in 0 until arr.length()) {
                    val opt = arr.getJSONObject(i)
                    add(DecisionOption(opt.getString("key"), opt.getString("label")))
                }
            }
        }

        val steps = buildList {
            val arr = json.optJSONArray("planSteps")
            if (arr != null) {
                for (i in 0 until arr.length()) {
                    val s = arr.getJSONObject(i)
                    add(
                        AiPlanStep(
                            num = s.optString("num", "①"),
                            title = s.getString("title"),
                            type = s.getString("type"),
                            typeColor = s.optString("typeColor", "var(--accent)"),
                            desc = s.optString("desc", ""),
                        ),
                    )
                }
            }
        }

        return GeneratedPlan(
            wishText = json.optString("wishText", fallbackWishText),
            durationText = json.optString("durationText", "预计 5 天完成"),
            decisionTitle = json.optString("decisionTitle", ""),
            decisionOptions = options,
            planSteps = steps,
            category = json.optString("category", "生活体验"),
            difficulty = json.optString("difficulty", "medium"),
            estimatedDays = json.optInt("estimatedDays", 5),
        )
    }

    private fun parseAnalysis(json: JSONObject): WishAnalysis {
        val constraints = buildList {
            val arr = json.optJSONArray("constraints")
            if (arr != null) for (i in 0 until arr.length()) add(arr.getString(i))
        }
        val preferences = buildList {
            val arr = json.optJSONArray("preferences")
            if (arr != null) for (i in 0 until arr.length()) add(arr.getString(i))
        }

        return WishAnalysis(
            intentType = json.optString("intentType", "execution"),
            confidence = json.optDouble("confidence", 0.5),
            executableAutomatically = json.optBoolean("executableAutomatically", false),
            needsFriendHelp = json.optBoolean("needsFriendHelp", false),
            needsCommunityHelp = json.optBoolean("needsCommunityHelp", false),
            goal = json.optString("goal", ""),
            constraints = constraints,
            preferences = preferences,
            timeframe = json.optString("timeframe").takeIf { it.isNotBlank() },
            budget = json.optString("budget").takeIf { it.isNotBlank() },
        )
    }

    companion object {
        private const val TAG = "AgentApi"
        // Android 模拟器访问宿主机 localhost 需要用 10.0.2.2
        // 真机调试需要用电脑的局域网 IP
        const val AI_SERVER_URL = "http://10.0.2.2:3100"
    }
}

sealed class AiPlanResult {
    data class Success(val plan: GeneratedPlan, val provider: String) : AiPlanResult()
    data class Error(val message: String) : AiPlanResult()
}

// ── 本地模板降级方案 — 对标 Web 端 generatePlanFromTemplate ──

private fun generateLocalTemplate(wishInput: String): GeneratedPlan {
    val input = wishInput.trim().lowercase()

    if (input.containsAny("海边", "海滩", "放松", "旅行", "度假")) {
        return GeneratedPlan(
            wishText = wishInput.trim(),
            durationText = "预计 3-5 天完成",
            decisionTitle = "AI 需要你决定：这次海边之行你更想要什么体验？",
            decisionOptions = listOf(
                DecisionOption("relax", "纯放松休闲"),
                DecisionOption("activity", "海上活动体验"),
                DecisionOption("photo", "拍照打卡风景"),
            ),
            planSteps = listOf(
                AiPlanStep("①", "筛选适合的海边目的地和住宿", "线上直出", "#4AADA0", "AI 自动搜索推荐"),
                AiPlanStep("②", "预订交通和酒店，准备物品清单", "资源助力", "#F5C842", "平台资源助力"),
                AiPlanStep("③", "找同行伙伴或当地向导推荐", "人群助力", "#c084fc", "AI匹配志同道合的旅友"),
                AiPlanStep("④", "开始海边放松之旅", "需你到场", "#f97316", "你本人享受旅程"),
            ),
            category = "生活体验",
            difficulty = "easy",
            estimatedDays = 4,
        )
    }

    if (input.containsAny("学习", "技能", "考试", "读书")) {
        return GeneratedPlan(
            wishText = wishInput.trim(),
            durationText = "预计 10-14 天完成",
            decisionTitle = "AI 需要你决定：你更偏好哪种学习模式？",
            decisionOptions = listOf(
                DecisionOption("self", "自主学习"),
                DecisionOption("course", "系统课程"),
                DecisionOption("mentor", "导师指导"),
            ),
            planSteps = listOf(
                AiPlanStep("①", "制定学习计划和时间安排", "线上直出", "#4AADA0", "AI 自动规划"),
                AiPlanStep("②", "整理学习资源和材料", "资源助力", "#F5C842", "平台资源助力"),
                AiPlanStep("③", "寻找学习伙伴或导师", "人群助力", "#c084fc", "AI匹配学习搭子"),
                AiPlanStep("④", "开始系统化学习", "需你到场", "#f97316", "你本人投入学习"),
            ),
            category = "学习成长",
            difficulty = "medium",
            estimatedDays = 12,
        )
    }

    if (input.containsAny("运动", "跑步", "健身", "锻炼")) {
        return GeneratedPlan(
            wishText = wishInput.trim(),
            durationText = "预计 7 天完成",
            decisionTitle = "AI 需要你决定：你更喜欢哪种运动节奏？",
            decisionOptions = listOf(
                DecisionOption("light", "轻松入门"),
                DecisionOption("regular", "规律训练"),
                DecisionOption("intense", "高强度挑战"),
            ),
            planSteps = listOf(
                AiPlanStep("①", "制定适合的运动计划", "线上直出", "#4AADA0", "AI 个性化规划"),
                AiPlanStep("②", "准备运动装备和场地", "资源助力", "#F5C842", "平台资源助力"),
                AiPlanStep("③", "寻找运动伙伴或教练", "人群助力", "#c084fc", "AI匹配运动搭子"),
                AiPlanStep("④", "开始规律运动", "需你到场", "#f97316", "你本人坚持锻炼"),
            ),
            category = "运动健康",
            difficulty = "easy",
            estimatedDays = 7,
        )
    }

    // 通用模板
    return GeneratedPlan(
        wishText = wishInput.trim(),
        durationText = "预计 5 天完成",
        decisionTitle = "AI 需要你决定：你更偏好哪种执行方式？",
        decisionOptions = listOf(
            DecisionOption("solo", "独自完成"),
            DecisionOption("partner", "寻找搭子"),
            DecisionOption("community", "社区协助"),
        ),
        planSteps = listOf(
            AiPlanStep("①", "分析需求并制定计划", "线上直出", "#4AADA0", "AI 智能分析"),
            AiPlanStep("②", "准备必要的资源和信息", "资源助力", "#F5C842", "平台资源助力"),
            AiPlanStep("③", "寻找合适的协助伙伴", "人群助力", "#c084fc", "AI匹配合适搭子"),
            AiPlanStep("④", "开始执行你的心愿", "需你到场", "#f97316", "你本人参与"),
        ),
        category = "生活体验",
        difficulty = "medium",
        estimatedDays = 5,
    )
}

private fun String.containsAny(vararg keywords: String): Boolean =
    keywords.any { this.contains(it) }
